### PR TITLE
Crossvalidation

### PR DESCRIPTION
**Pull request by KEggensperger**
This PR adds a crossvalidation version of logistic regression and fixes some issues with the random seed.

### CODE REVIEW

**Comment by KEggensperger** [comment]
This PR is now ready for reviewing
@aaronkl @mfeurer

File: .travis.yml [hunk 1]
```diff
@@ -33,8 +33,8 @@ install:
   - cat requirements.txt | grep -v '#' | xargs -n 1 -L 1 pip install
   - cat optional-requirements.txt | grep -v '#' | xargs -n 1 -L 1 pip install
   # Install auto-sklearn
-  - curl https://raw.githubusercontent.com/automl/auto-sklearn/development/requirements.txt | xargs -n 1 -L 1 pip install
-  - pip install git+https://github.com/automl/auto-sklearn@development
+  #- curl https://raw.githubusercontent.com/automl/auto-sklearn/development/requirements.txt | xargs -n 1 -L 1 pip install
```
Comment (mfeurer) [inline .travis.yml]: Does this imply that `auto-sklearn` should be an optional dependency?

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+def mini_wrapper(configuration, bench, **kwargs):
+    """ mini wrapper to organize arguments """
+    res_dict = bench.objective_function(configuration,
+                                        fold=kwargs["instance"],
+                                        rng=kwargs["seed"])
+    return res_dict["function_value"]
+
+logReg = logistic_regression.LogisticRegression10CVOnMnist(rng=10)
+
+logger = logging.getLogger("Optimizer")  # Enable to show Debug outputs
```
Comment (mfeurer) [inline examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: I don't see why one needs the `logger` variable.

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ta_call = functools.partial(mini_wrapper, bench=logReg)
+taf = ExecuteTAFuncDict(ta_call)
+
+# SMAC scenario object
+scenario = Scenario({"run_obj": "quality",
+                     "runcount-limit": 5,
+                     "cs": cs,
+                     "deterministic": "true",
+                     "instances": list([str(i) for i in range(10)]), # train_inst_fh.name,
+                     "test_instances": ["10", ]
```
Comment (mfeurer) [inline examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: You don't need the comma in a list.

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                     })
+
+# Evaluate default config as a test
```
Comment (mfeurer) [inline examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: I don't understand the comment.

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+# Use SMAC to optimize function
+smac = SMAC(scenario=scenario, rng=1,
+            tae_runner=taf)
+try:
+    incumbent = smac.optimize()
+finally:
+    incumbent = smac.solver.incumbent
+
+# Then evaluate final configuration
+inc_value = taf.run(incumbent)[1]
```
Comment (mfeurer) [inline examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: Shouldn't that run on the test set?

File: hpolib/benchmarks/ml/logistic_regression.py [hunk 6]
```diff
@@ -4,46 +4,55 @@
...
-        pass
+    def get_data(self):
+        raise NotImplementedError("Do not use this benchmark as this is only "
+                                  "a skeleton for further implementations.")
 
     @AbstractBenchmark._check_configuration
     @AbstractBenchmark._configuration_as_array
     def objective_function(self, x, dataset_fraction=1, **kwargs):
+        # Shuffle training data
+        rng = kwargs.get("rng", None)
```
Comment (mfeurer) [inline hpolib/benchmarks/ml/logistic_regression.py]: `None` is the default return value of `dict.get()` and not necessary.

File: hpolib/benchmarks/ml/logistic_regression.py [hunk 7]
```diff
@@ -53,51 +62,41 @@ def objective_function(self, x, dataset_fraction=1, **kwargs):
...
 
-        start_time = time.time()
+        rng = kwargs.get("rng", None)
+        self.rng = rng_helper.get_rng(rng=rng, self_rng=self.rng)
 
-        learning_rate = np.float32(10 ** x[0])
-        l2_reg = np.float32(x[1])
-        batch_size = np.int32(x[2])
-        dropout_rate = np.float32(x[3])
+        # if rng was not not, set rng for lasagne
```
Comment (mfeurer) [inline hpolib/benchmarks/ml/logistic_regression.py]: two `not` in the comment

File: hpolib/benchmarks/ml/logistic_regression.py [hunk 8]
```diff
@@ -120,6 +119,32 @@ def get_meta_information():
                                "year = {2016}}"]
                 }
 
+    def _train_model(self, config, train, train_targets, valid, valid_targets):
+        """ helper method that accepts data and configuration and returns
+        lc_curve, cost_curve
+        :param config: list
```
Comment (mfeurer) [inline hpolib/benchmarks/ml/logistic_regression.py]: shouldn't this be in numpy format?

File: hpolib/benchmarks/ml/logistic_regression.py [hunk 9]
```diff
@@ -120,6 +119,32 @@ def get_meta_information():
...
+            hyperparameters [learning_rate, l2_reg, batch_size, dropout_rate]
+        :param train: array
+        :param train_targets: list
+        :param valid: array
+        :param valid_targets: list
+        :return: lc_curve, cost_curve
+        """
+        learning_rate = float(10 ** config[0])
```
Comment (mfeurer) [inline hpolib/benchmarks/ml/logistic_regression.py]: I assume there was a reason why the values were casted to `np.float32` before this change request.

File: hpolib/util/rng_helper.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+
+
+def get_rng(rng=None, self_rng=None):
+    """ helper function to obtain RandomState.
+    returns RandomState created from rng
+    if rng is None returns self_rng if RandomState
+    if self_rng is None initializes RandomState at random
+
+    :param rng: int or RandomState
```
Comment (mfeurer) [inline hpolib/util/rng_helper.py:10]: Shouldn't this be in numpy docstyle format?

File: hpolib/util/rng_helper.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        return create_rng(rng)
+    elif rng is None and self_rng is not None:
+        return create_rng(self_rng)
+    else:
+        return np.random.RandomState()
+
+
+def create_rng(rng):
+    """ helper to create rng from RandomState or int
+    :param rng: int or RandomState
```
Comment (mfeurer) [inline hpolib/util/rng_helper.py:25]: Shouldn't this be in numpy docstyle format?

File: setup.py [hunk 12]
```diff
@@ -4,19 +4,12 @@
...
-
-
-
 setuptools.setup(
     name='hpolib2',
     description='Automated machine learning.',
     version=version,
     packages=setuptools.find_packages(exclude=['test']),
-    install_requires=requirements,
+    install_requires=[],
```
Comment (mfeurer) [inline setup.py:12]: why does this PR remove the requirements?

**Review by mfeurer — CHANGES_REQUESTED** [review]

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py (hunk 4 above)
Comment (KEggensperger) [reply examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: Agreed and removed

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py (hunk 5 above)
Comment (KEggensperger) [reply examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: Agreed and changed

File: examples/SMAC3_on_LogisticRegression10CVOnMnist.py (hunk 2 above)
Comment (KEggensperger) [reply examples/SMAC3_on_LogisticRegression10CVOnMnist.py]: Agreed and removed

File: .travis.yml (hunk 1 above)
Comment (KEggensperger) [reply .travis.yml]: No, auto-sklearn is installed in l33 as a mandatory requirement

File: hpolib/benchmarks/ml/logistic_regression.py (hunk 8 above)
Comment (KEggensperger) [reply hpolib/benchmarks/ml/logistic_regression.py]: Not my code, but changed it

File: hpolib/benchmarks/ml/logistic_regression.py (hunk 7 above)
Comment (KEggensperger) [reply hpolib/benchmarks/ml/logistic_regression.py]: Agreed

File: hpolib/benchmarks/ml/logistic_regression.py (hunk 9 above)
Comment (KEggensperger) [reply hpolib/benchmarks/ml/logistic_regression.py]: I don't know. @aaronkl: Is there a reason for this?

File: hpolib/util/rng_helper.py (hunk 10 above)
Comment (KEggensperger) [reply hpolib/util/rng_helper.py:10]: Agreed and changed it

File: setup.py (hunk 12 above)
Comment (KEggensperger) [reply setup.py:12]: Because the documentation recommends to install them independently and our current requirements.txt contains comments which breaks our current setup.py

**Review by mfeurer — APPROVED** [review]
